src/vmm/src/builder.rs

```diff
@@ -655,75 +650,47 @@ fn attach_mmio_device(
         .device_type();
     let cmdline = &mut vmm.kernel_cmdline;
 
-    vmm.mmio_device_manager.register_mmio_device(
-        vmm.vm.fd(),
-        device,
-        cmdline,
-        type_id,
-        id.as_str(),
-    )?;
+    vmm.mmio_device_manager
+        .register_mmio_device(vmm.vm.fd(), device, cmdline, type_id, id)?;
 
     Ok(())
 }
 
 fn attach_block_devices(
     vmm: &mut Vmm,
-    blocks: &BlockDeviceConfigs,
+    blocks: &BlockDevices,
     event_manager: &mut EventManager,
 ) -> std::result::Result<(), StartMicrovmError> {
     use self::StartMicrovmError::*;
 
-    for drive_config in blocks.config_list.iter() {
-        // Add the block device from file.
-        let block_file = OpenOptions::new()
-            .read(true)
-            .write(!drive_config.is_read_only)
-            .open(&drive_config.path_on_host)
-            .map_err(OpenBlockDevice)?;
-
-        if drive_config.is_root_device {
+    for (idx, block) in blocks.list.iter().enumerate() {
+        if blocks.has_root_block && idx == 0 {
             let kernel_cmdline = &mut vmm.kernel_cmdline;
 
-            kernel_cmdline.insert_str(if let Some(partuuid) = &drive_config.partuuid {
+            let locked = block.lock().unwrap();
+            kernel_cmdline.insert_str(if let Some(partuuid) = locked.partuuid() {
                 format!("root=PARTUUID={}", partuuid)
             } else {
                 // If no PARTUUID was specified for the root device, try with the /dev/vda.
                 "root=/dev/vda".to_string()
             })?;
 
-            let flags = if drive_config.is_read_only {
-                "ro"
-            } else {
-                "rw"
-            };
+            let flags = if locked.is_read_only() { "ro" } else { "rw" };
 
             kernel_cmdline.insert_str(flags)?;
         }
 
-        let rate_limiter = drive_config
-            .rate_limiter
-            .map(vmm_config::RateLimiterConfig::try_into)
-            .transpose()
-            .map_err(CreateRateLimiter)?;
-
-        let block_device = Arc::new(Mutex::new(
-            devices::virtio::Block::new(
-                block_file,
-                drive_config.is_read_only,
-                rate_limiter.unwrap_or_default(),
-            )
-            .map_err(CreateBlockDevice)?,
-        ));
-
         event_manager
-            .add_subscriber(block_device.clone())
-            .map_err(StartMicrovmError::RegisterEvent)?;
+            .add_subscriber(block.clone())
+            .map_err(RegisterEvent)?;
 
+        let id = block.lock().unwrap().id().clone();
+        // The device mutex mustn't be locked here otherwise it will deadlock.
         attach_mmio_device(
             vmm,
-            drive_config.drive_id.clone(),
-            MmioTransport::new(vmm.guest_memory().clone(), block_device.clone())
-                .map_err(CreateBlockDevice)?,
+            id,
```

Comment:
I would avoid adding the id parameter here. Are there any options of fetching it inside the function ?